locator/src/locator.rs

```diff
@@ -98,73 +97,146 @@ pub enum LoadError {
+    // Standard interval between refresh attempts.
+    refresh_interval: std::time::Duration,
+    // Minimum duration between refresh attempts.
+    min_refresh_interval: std::time::Duration,
+    // Channel to send commands to the loader task.
+    tx: mpsc::Sender<Command>,
 }
 
 impl OrgToCell {
     pub fn new(
         control_plane_url: String,
         backup_routes: Arc<dyn BackupRouteProvider + Send + Sync>,
         locality_to_default_cell: Option<HashMap<String, String>>,
+        tx: mpsc::Sender<Command>,
     ) -> Self {
-        OrgToCell {
-            control_plane: ControlPlane::new(control_plane_url),
-            locality_to_default_cell: locality_to_default_cell.unwrap_or_default(),
-            data: RwLock::new(RouteData {
+        let data = RouteDataWithTimestamp {
+            data: RouteData {
                 org_to_cell: HashMap::new(),
                 last_cursor: "".into(),
                 cells: HashMap::new(),
-            }),
+            },
+            last_updated: None,
+        };
+
+        OrgToCell {
+            control_plane: ControlPlane::new(control_plane_url),
+            locality_to_default_cell: locality_to_default_cell.unwrap_or_default(),
+            data: RwLock::new(data),
+            negative_cache: NegativeCache::new(),
             update_lock: Semaphore::new(1),
             ready: AtomicBool::new(false),
             backup_routes,
+            refresh_interval: std::time::Duration::from_secs(60),
+            min_refresh_interval: std::time::Duration::from_secs(1),
+            tx,
         }
     }
 
-    pub fn lookup(&self, org_id: &str, locality: Option<&str>) -> Result<String, LocatorError> {
+    pub async fn lookup(
+        &self,
+        org_id: &str,
+        locality: Option<&str>,
+    ) -> Result<String, LocatorError> {
         // Looks up the cell for a given organization ID and locality.
-        // Returns an `Option<Cell>` if found, or `None` if not found.
+        // Returns `Ok(Cell)` if found, or a default applies.
         // Returns an error if locality is passed and the org_id/locality pair is not valid.
         // Or if a locality is passed but no defualt cell is found for that locality
         if !self.ready.load(Ordering::Relaxed) {
             return Err(LocatorError::NotReady);
         }
 
-        let read_guard = self.data.read();
-        let cell_id = read_guard
-            .org_to_cell
-            .get(org_id)
-            .or_else(|| {
-                if let Some(loc) = locality {
-                    self.locality_to_default_cell.get(loc)
-                } else {
-                    None
+        let start_lookup = Instant::now();
+
+        // Try to get cell_id from cache. The read lock is only held briefly.
+        let cell_id = {
+            let read_guard = self.data.read();
+            read_guard.data.org_to_cell.get(org_id).cloned()
+        };
+
+        // If not found in cache, try refreshing
```

Comment:
this file had to be restructured a bit so we don't hold onto the read lock for the whole function, as that would cause a deadlock if incremental load needs to change the data.